core/src/price_estimation/orderbook_based.rs

```diff
@@ -0,0 +1,149 @@
+use super::price_source::PriceSource;
+use crate::models::{BatchId, TokenId};
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use pricegraph::{Orderbook, TokenPair};
+use primitive_types::U256;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+pub struct PricegraphEstimator {
+    orderbook_reader: Arc<dyn StableXOrderBookReading>,
+}
+
+const ONE_OWL: f64 = 1_000_000_000_000_000_000.0;
+
+impl PriceSource for PricegraphEstimator {
+    fn get_prices<'b>(
+        &'b self,
+        tokens: &'b [TokenId],
+    ) -> BoxFuture<'b, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let batch = BatchId::currently_being_solved(SystemTime::now())?;
+            let (account_state, orders) =
+                self.orderbook_reader.get_auction_data(batch.into()).await?;
+            let orderbook = Orderbook::from_elements(orders.iter().map(|order| {
+                order.to_element(U256(
+                    account_state
+                        .read_balance(order.sell_token, order.account_id)
+                        .0,
+                ))
+            }));
+            Ok(self.estimate_prices(tokens, orderbook))
+        }
+        .boxed()
+    }
+}
+
+// Private trait to facilitate testing this module
+#[cfg_attr(test, mockall::automock)]
+trait Pricegraph {
+    fn estimate_price(&self, pair: TokenPair, volume: f64) -> Option<f64>;
```

Comment:
Nice! I had saw it last night with the `&mut self` and wanted to suggest changing it to `&self` because it needs to be cloned anyway.